tso-server/src/main/java/com/yahoo/omid/tso/PersistenceProcessorImpl.java

```diff
@@ -15,345 +15,216 @@
         this.panicker = panicker;
 
-        LOG.info("Creating the persist processor with batch size {}, and timeout {}ms",
-                config.getMaxBatchSize(), config.getBatchPersistTimeoutMS());
-
-        flushTimer = metrics.timer(name("tso", "persist", "flush"));
-        batchSizeHistogram = metrics.histogram(name("tso", "persist", "batchsize"));
-        timeoutMeter = metrics.meter(name("tso", "persist", "timeout"));
-
-        // FIXME consider putting something more like a phased strategy here to avoid
-        // all the syscalls
-        final TimeoutBlockingWaitStrategy timeoutStrategy
-                = new TimeoutBlockingWaitStrategy(config.getBatchPersistTimeoutMS(), TimeUnit.MILLISECONDS);
+        batchIDCnt = 0L;
 
-        persistRing = RingBuffer.createSingleProducer(
-                PersistEvent.EVENT_FACTORY, 1 << 20, timeoutStrategy); // 2^20 entries in ringbuffer
-        SequenceBarrier persistSequenceBarrier = persistRing.newBarrier();
-        BatchEventProcessor<PersistEvent> persistProcessor = new BatchEventProcessor<>(
-                persistRing,
-                persistSequenceBarrier,
-                this);
-        persistRing.addGatingSequences(persistProcessor.getSequence());
-        persistProcessor.setExceptionHandler(new FatalExceptionHandler(panicker));
+        batchPool = new BatchPool(config, batch);
 
-        ExecutorService persistExec = Executors.newSingleThreadExecutor(
-                new ThreadFactoryBuilder().setNameFormat("persist-%d").build());
-        persistExec.submit(persistProcessor);
+        this.batch = batchPool.getNextEmptyBatch();
 
-    }
+        timeoutMeter = metrics.meter(name("tso", "persist", "timeout"));
 
-    @Override
-    public void onEvent(PersistEvent event, long sequence, boolean endOfBatch) throws Exception {
-
-        switch (event.getType()) {
-            case COMMIT:
-                event.getMonCtx().timerStart("commitPersistProcessor");
-                // TODO: What happens when the IOException is thrown?
-                writer.addCommittedTransaction(event.getStartTimestamp(), event.getCommitTimestamp());
-                batch.addCommit(event.getStartTimestamp(), event.getCommitTimestamp(), event.getChannel(),
-                                event.getMonCtx());
-                break;
-            case ABORT:
-                sendAbortOrIdentifyFalsePositive(event.getStartTimestamp(), event.isRetry(), event.getChannel(),
-                                                 event.getMonCtx());
-                break;
-            case TIMESTAMP:
-                event.getMonCtx().timerStart("timestampPersistProcessor");
-                batch.addTimestamp(event.getStartTimestamp(), event.getChannel(), event.getMonCtx());
-                break;
-        }
+        persistRing = RingBuffer.<PersistBatchEvent>createSingleProducer(
+                PersistBatchEvent.EVENT_FACTORY, 1 << 20, new BusySpinWaitStrategy());
```

Comment:
I used BusySpinWait because this is what all the other processors use. As you wrote this is since Omid is Throughput oriented.
I agree that making this configurable for all the processor, and not only for this one, would be great. However, IMHO this should be in a different commit.